document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/transformer/MongoSelectionsUpdatingTransformation.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * <p>For example, this transformation converts the aggregate pipeline <code>
+ *   [
+ *      {
+ *        "$group": {
+ *          "_id": {
+ *            "item": "$item",
+ *            "price": "$props.price"
+ *          },
+ *          "num_brands": {
+ *            "$distinctCount": "$brand"
+ *          }
+ *        }
+ *      },
+ *      {
+ *        "$project": {
+ *          "item": 1
+ *        }
+ *      }
+ *   ]
+ * </code> into <code>
+ *   [
+ *      {
+ *        "$group": {
+ *          "_id": {
+ *            "item": "$item",
+ *            "price": "$props.price"
+ *          },
+ *          "num_brands": {
+ *            "$distinct": "$brand"
+ *          }
+ *        }
+ *      },
+ *      {
+ *        "$project": {
+ *          "item": "$_id.item"
+ *        }
+ *      }
+ *   ]
+ * </code> since "item" appears in projection as well as grouping and "$distinctCount" is not
+ * supported
+ */
+final class MongoSelectionsUpdatingTransformation implements SelectingExpressionVisitor {
+  private static final Function<AggregateExpression, AggregateExpression> COUNT_HANDLER =
+      expression -> AggregateExpression.of(SUM, ConstantExpression.of(1));
+
+  private static final Function<AggregateExpression, AggregateExpression> DISTINCT_COUNT_HANDLER =
+      expression -> AggregateExpression.of(DISTINCT, expression.getExpression());
+
+  private static final Map<AggregationOperator, Function<AggregateExpression, AggregateExpression>>
+      AGGREGATION_SUBSTITUTE_MAP =
+          unmodifiableMap(
+              new EnumMap<>(AggregationOperator.class) {
+                {
+                  put(DISTINCT_COUNT, DISTINCT_COUNT_HANDLER);
+                  put(COUNT, COUNT_HANDLER);
+                }
+              });
+
+  private final List<GroupingExpression> aggregations;
+  private final SelectionSpec source;
+  private final Map<Integer, GroupingExpression> aggregationMap;
+
+  MongoSelectionsUpdatingTransformation(
+      List<GroupingExpression> aggregations, SelectionSpec source) {
+    this.aggregations = aggregations;
+    this.source = source;
+    this.aggregationMap = getAggregationMap();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final AggregateExpression expression) {
+    return substitute(expression);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final ConstantExpression expression) {
+    return source;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final FunctionExpression expression) {
+    return source;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final IdentifierExpression expression) {
+    GroupingExpression matchingGroup = aggregationMap.get(expression.hashCode());
+    if (!expression.equals(matchingGroup)) {
+      return source;
+    }
+
+    String key = expression.getName();
+    String identifier = MongoCollection.ID_KEY + "." + key;
```

Comment:
nit : extract into constant

Reply:
You mean, make identifier as a final variable?